# this code is intended to implement FIFO circular buffer in 2 classes the first being the following CircularBuffer class
### the first implementation will be done using a normal array

In [1]:
class CircularBuffer:
    def __init__(self, buffSize):
        self.buffsize=buffSize
        self.data=[]
    class FullBuffer:
        def add(self, x):
            self.data[self.curpos]=x
            self.curpos=(self.curpos+1)%self.buffsize
        
        def get(self):
            return self.data[self.curpos:]+self.data[:self.curpos] ## return the results by deviding the array in half 
        ## we return the first half until the current position concatinated with the rest of the array 
        ## we do that to avoid having to shift the elements in order to achieve the right order putting the element
        ## at the current position at the top of the buffer 


    def add(self, x):
        self.data.append(x)
        if len(self.data)>=self.buffsize:
            self.curpos=0
            ## check when the buffer is full 

            self.__class__=self.FullBuffer
            ## if full, move to the full buffer implementation 
            
            
    def get(self):
        return self.data
        

### the second implementation will be done using a deque

In [ ]:


import collections
class CircularBufferDeque:
    def __init__(self, buffSize):
        self.dq=collections.deque(maxlen=buffSize)
    def add(self, x):
        self.dq.append(x)
    def get(self):
        return self.dq

### to compare the two implementations in terms of time we will take advantage of the timeIt library

In [39]:
import timeit
def main():
    
    
    bufferSize=100
    numIterations =1e5
    resultAdd=[]
    resultGet=[]
    circularBuffers=[CircularBuffer(bufferSize), CircularBufferDeque(bufferSize)]
    implementations=["circular buffer with array", "circular buffer with deque"]
    print(circularBuffers)
    for i, cb in enumerate(circularBuffers):
        timer = timeit.Timer(lambda: [circularBuffers[i].add(j) for j in range(bufferSize)])
        elapsed_time = timer.timeit(int(numIterations))
        resultAdd.append(elapsed_time)
        timer2=timeit.Timer(lambda: circularBuffers[i].get())
        elapsed_time2 = timer2.timeit(int(numIterations))
        
        resultGet.append(elapsed_time2)
    for i, res in enumerate(resultAdd):
        print(' time for add opperation for implementation {} the time measured is: {:.2f}s, ({:.2f} per iteration) for the number of iterations {} '
              .format( implementations[i],res, res/numIterations*1e4, numIterations)
                                                                                                        )
    print("##########")
    for i, res in enumerate(resultGet):
        print(' time for get opperation for implementation {} the time measured is: {:.2f}s, ({:.2f} per iteration) for the number of iterations {} '
              .format( implementations[i],res, res/numIterations*1e4, numIterations)
                                                                                                        )
        print("##########")
                                       
        
    
    return 
if __name__=="__main__":
    main()

[<__main__.CircularBuffer object at 0x00000219404FCAC0>, <__main__.CircularBufferDeque object at 0x000002194049E3E0>]
 time for add opperation for implementation circular buffer with array the time measured is: 4.16s, (0.42 per iteration) for the number of iterations 100000.0 
 time for add opperation for implementation circular buffer with deque the time measured is: 2.54s, (0.25 per iteration) for the number of iterations 100000.0 
##########
 time for get opperation for implementation circular buffer with array the time measured is: 0.11s, (0.01 per iteration) for the number of iterations 100000.0 
##########
 time for get opperation for implementation circular buffer with deque the time measured is: 0.03s, (0.00 per iteration) for the number of iterations 100000.0 
##########


## analyzing the results
### in terms of time
for the metrics used which are 100 000 iterations 
the times for both the add and the get operation are significantly better in the case of using a deque for the implementation, Given the fact that we got 2.54 and 0.03 seconds for the total time for the get and add operations respectively. Compared to the 4.16 and 0.11 seconds that we got when implementing the methods using an array.
### in terms of readability

Implementing the methods using deque is simpler and easier to understand than doing that using an array
### underlying reasons for the results recieved


Using deque to implement the circular buffer is the more natural approach for the following reasons:
Operations in deque are of complexity O(1), due to it's two pointer implementation with one pointer at the end and the other one on the front of the queue. In this use case we only need the push and pop operations since we are not concerned with the get element at a certain index method, which in either way matches the complexity of it being implemented with an array and that is O(n). 
on the other hand, implementing the methods using the simple array approach gives a O(n) time complexity in the case of a full buffer and that is because of the need to insert an element at a certain index. 
So to sum up, as mentioned before I recommend using a deque for this use case. 

